archive/imo/imo2022_q2.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2022 boboquack. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: boboquack.
+-/
+import data.real.basic
+import tactic
+
+/-!
+# IMO 2022 Q2
+
+Let `ℝ+` denote the set of positive real numbers. Find all functions `f : ℝ+ → ℝ+` 
+such that for each `x ∈ ℝ+`, there is exactly one `y ∈ ℝ+` satisfying
+`x f(y) + y f(x) ≤ 2`.
+
+## Sketch of solution
+
+`f(x) := 1/x` is the only solution. We first verify this is a solution, always
+taking `x=y`. To show this is the only solution, we first prove that the given
+identity implies `x=y` because otherwise we can derive `f(x) ≥ x⁻¹` and `f(y) ≥ y⁻¹`,
+swiftly leading to contradiction. This then shows `f(x) ≤ x⁻¹` for all `x`, but
+if this inequality is strict anywhere we obtain contradiction in the uniqueness
+of `y` by taking `y` close enough to `x` (specified explicitly in this
+formalisation), hence `f(x) := 1/x` is the only solution.
+
+We make repeated use of the `recips_sum_two` lemma, which is a trivial consequence
+of the AM-GM inequality but which can also be obtained from square factorisation.
+For ease of manipulation, we take `0 < (each variable)` as hypotheses rather than
+proving known facts about the reals translate to `ℝ+`.
+-/
+
+lemma recips_sum_two {x y : ℝ} (hx : 0 < x) (hy : 0 < y)
+  (hi : x * y⁻¹ + y * x⁻¹ ≤ 2) : x = y := 
+begin 
+  -- clear denominators and expand
+  rw ← (mul_le_mul_right (mul_pos hx hy)) at hi,
+  rw (show (x * y⁻¹ + y * x⁻¹) * (x * y) = x * x + y * y, by 
+  { field_simp [hx.ne', hy.ne'],
+    left, 
+    ring,}) at hi,
+  -- the conclusion is now trivial from `(x - y)² = 0` implying `x=y`
+  nlinarith,
+end
+
+theorem imo2022Q2 (f: ℝ → ℝ) : 
+  (∀ {x : ℝ}, 0 < x → (0 < f x ∧ ∃! y : ℝ, 0 < y ∧ x * f y + y * f x ≤ 2) )
+  ↔ ∀ {x : ℝ}, 0 < x → f x = x⁻¹ := 
```

Comment:
What are your thoughts on writing this as
```suggestion
notation `ℝ+` := set.Ioi (0 : ℝ) -- or similar

theorem imo2022Q2 (f : ℝ+ → ℝ+) : 
  (∀ x : ℝ+, ∃! y : ℝ+, (x * f y + y * f x : ℝ) ≤ 2) ↔ ∀ x : ℝ+, (f x : ℝ) = x⁻¹ := 
```
All the operations are still happening in `ℝ`, so this shouldn't be much harder to work with; but it much more closely resembles the original statement.

Reply:
Good idea. I wasn't sure how best to handle ℝ+: dealing with {x : ℝ | 0 < x} was painful but this seems to have worked reasonably well.

(I think I may have done bad things with Github in the process - not entirely sure how I go ensuring my pull request doesn't carry old versions of mathlib files while not clogging the history with extra merges.)